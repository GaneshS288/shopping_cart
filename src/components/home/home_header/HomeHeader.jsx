import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";

function HomeHeader({ userName, cartItemCount }) {
  return (
    <div>
      <NavLink to={"/home"}>BlazingCart</NavLink>

      <div>
        <button>
          <NavLink to={"/home/cart"}>
            <div>
              <img src="" alt="cart-icon" />
              {cartItemCount > 0 ? <span data-testid="cartItemCount">{cartItemCount}</span> : null}
            </div>
          </NavLink>
        </button>

        <div>
          <img src="" alt="user avatar" />
          <p>{userName}</p>
        </div>
      </div>
    </div>
  );
}

HomeHeader.propTypes = {
  userName: PropTypes.string.isRequired,
  cartItemCount: PropTypes.number.isRequired,
};

export default HomeHeader;
