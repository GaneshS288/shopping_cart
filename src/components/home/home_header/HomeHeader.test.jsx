import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import HomeHeader from "./HomeHeader";

function DummyComponent() {
  return <p>I am dummy</p>;
}

function testSetup(cartItemCount = 2) {
  const username = "Ganesh";

  const routes = [
    {
      path: "/home",
      element: (
        <HomeHeader
          userName={username}
          cartItemCount={cartItemCount}
        ></HomeHeader>
      ),
    },

    {
      path: "/home/cart",
      element: <DummyComponent></DummyComponent>,
    },
  ];

  const router = createMemoryRouter(routes, {
    initialEntries: ["/", "/home"],
    initialIndex: 1,
  });

  const user = userEvent.setup();

  return { router, user };
}

describe("Home header", () => {
  it("matches snapshot", () => {
    const { router } = testSetup();

    const { container } = render(
      <RouterProvider router={router}></RouterProvider>
    );

    screen.debug();

    expect(container).toMatchSnapshot();
  });

  it("renders the header", () => {
    const { router } = testSetup();

    render(<RouterProvider router={router}></RouterProvider>);

    const [homeLink, cartLink] = screen.getAllByRole("link");

    expect(homeLink.textContent).toMatch("BlazingCart");
    expect(cartLink).toBeInTheDocument();
  });

  it("renders name and cartItemCount props", () => {
    const { router } = testSetup();

    render(<RouterProvider router={router}></RouterProvider>);

    const userNamePara = screen.getByRole("paragraph");
    const cartItemCount = screen.getByTestId("cartItemCount");
    expect(userNamePara.textContent).toMatch("Ganesh");
    expect(cartItemCount.textContent).toMatch("2");
  });

  it("doesn't render cart item count if count is below 1", () => {
    const { router } = testSetup(0);

    render(<RouterProvider router={router}></RouterProvider>);

    const cartItemCount = screen.queryByTestId("cartItemCount");
    expect(cartItemCount).not.toBeInTheDocument();
  });

  it("navigates to cart", async () => {
    const { router, user } = testSetup();

    render(<RouterProvider router={router}></RouterProvider>);

    const cartLink = screen.getAllByRole("link")[1];

    await user.click(cartLink);
    
    expect(screen.getByRole("paragraph").textContent).toMatch("I am dummy");
  });
});
