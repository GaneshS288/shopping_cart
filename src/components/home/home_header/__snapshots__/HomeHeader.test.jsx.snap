// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home header > matches snapshot 1`] = `
<div>
  <div>
    <a
      aria-current="page"
      class="active"
      data-discover="true"
      href="/home"
    >
      BlazingCart
    </a>
    <div>
      <button>
        <a
          class=""
          data-discover="true"
          href="/home/cart"
        >
          <div>
            <img
              alt="cart-icon"
              src=""
            />
            <span
              data-testid="cartItemCount"
            >
              2
            </span>
          </div>
        </a>
      </button>
      <div>
        <img
          alt="user avatar"
          src=""
        />
        <p>
          Ganesh
        </p>
      </div>
    </div>
  </div>
</div>
`;
